File: frontend/src/types/Node.ts
export class Node {
  x: number;
  y: number;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  subtract(other: Node): Node {
    if (!(other instanceof Node)) {
      throw new TypeError("Argument must be an instance of Node");
    }
    return new Node(this.x - other.x, this.y - other.y);
  }

  add(other: Node): Node {
    if (!(other instanceof Node)) {
      throw new TypeError("Argument must be an instance of Node");
    }
    return new Node(this.x + other.x, this.y + other.y);
  }

  copy(): Node {
    return new Node(this.x, this.y);
  }

  distance(other: Node): number {
    if (!(other instanceof Node)) {
      throw new TypeError("Argument must be an instance of Node");
    }
    return Math.sqrt(Math.pow(this.x - other.x, 2) + Math.pow(this.y - other.y, 2));
  }

  lessThan(other: Node): boolean {
    return this.x < other.x || (this.x === other.x && this.y < other.y);
  }

  greaterThan(other: Node): boolean {
    return this.x > other.x || (this.x === other.x && this.y > other.y);
  }

  toString(): string {
    return `(${this.x}, ${this.y})`;
  }

  equals(other: Node): boolean {
    return this.x === other.x && this.y === other.y;
  }

  hash(): string {
    return `${this.x},${this.y}`;
  }
}

File: python - prototype/src/Node.py
import numpy as np


class Node:
  def __init__(self, x, y):
    self.x = x
    self.y = y
    
  def __sub__(self, other):
    "substracting Points"
    if not isinstance(other, Node):
      raise TypeError
    
    return Node(self.x - other.x, self.y - other.y)
  
  def __add__(self, other):
    "adding Points"
    if not isinstance(other, Node):
      raise TypeError
    
    return Node(self.x + other.x, self.y + other.y)
  
  def copy(self):
    return Node(self.x, self.y)
  
  def distance(self, other) -> float:
    "distance of two Points"
    if not isinstance(other, Node):
      raise TypeError
    
    return np.sqrt((self.x - other.x) ** 2 + (self.y - other.y) ** 2)
  
  def __lt__(self, other):
    return self.x < other.x or (self.x == other.x and self.y < other.y)
  
  def __gt__(self, other):
    return self.x > other.x or (self.x == other.x and self.y > other.y)
  
  def __str__(self):
    return f"({self.x}, {self.y})"
  
  def __repr__(self): 
    return str(self)
  
  def __eq__(self, other):
    return self.x == other.x and self.y == other.y
  
  def __hash__(self):
    return hash((self.x, self.y))